api/src/main/java/com/nicknathanjustin/streamercontracts/votes/VoteServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+                    .build();
+            voteModelRepository.save(voteModel);
+        }
+    }
+
+    @Override
+    public boolean isVotingComplete(@NonNull final ContractModel contractModel) {
+        validateContract(contractModel);
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        final Timestamp now = new Timestamp(System.currentTimeMillis());
+        return (proposerVote != null && streamerVote != null) ||
+               (proposerMarkedContractCompleted(contractModel)) ||
+               (streamerMarkedContractFailed(contractModel)) ||
+               (now.after(contractModel.getExpiresAt()));
+    }
+
+    @Override
+    public boolean wasContractCompleted(@NonNull final ContractModel contractModel) {
+        if (!isVotingComplete(contractModel)) {
+            throw new IllegalArgumentException("Attempted to check if contract was completed before voting was finished. ContractId: " + contractModel.getId());
+        }
+
+        final UUID contractId = contractModel.getId();
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getProposer().getId());
+        final VoteModel streamerVote = voteModelRepository.findByContractIdAndVoterId(contractId, contractModel.getStreamer().getId());
+        if (proposerMarkedContractCompleted(contractModel)) {
+            //Proposer has marked contract as completed. Release funds without reading Streamer's vote.
+            return true;
+        } else if (streamerMarkedContractFailed(contractModel)) {
+            //Streamer has marked contract as failed. Return funds without reading Proposers's vote.
+            return false;
+        } else if(proposerVote == null && streamerVote == null) {
+            //Neither proposer or streamer voted. Release funds if contract was accepted. Return funds otherwise.
+            return contractModel.isAccepted();
+        } else if(proposerVote != null) {
+            //Streamer did not vote. Take Proposer's vote as the final outcome
+            return proposerVote.isViewerFlaggedComplete();
+        } else if(streamerVote != null){
+            //Proposer did not vote. Take Streamer's vote as the final outcome
+            return streamerVote.isViewerFlaggedComplete();
+        }
+        //Default case is to consider the contract completed and release money to the streamer.
+        return true;
+    }
+
+    private boolean proposerMarkedContractCompleted(@NonNull final ContractModel contractModel) {
+        final VoteModel proposerVote = voteModelRepository.findByContractIdAndVoterId(contractModel.getId(), contractModel.getProposer().getId());
```

Comment:
Nit: this.voteModelRepository?

Reply:
We havent been using the this.* notation in our application. I think its unnecessary but we can discuss on the next call what pattern we should embrace moving forward. Not fixing for now.